### DISCUSSION TITLE
post/2021-03-06-grpc-go-discovery-lb/

### DISCUSSION THREAD

**Discussion by giscus** [zcong1993/blog#59] [Announcements]
# post/2021-03-06-grpc-go-discovery-lb/

gRPC 是 Google 开源的一款高性能, 支持多种语言的 RPC 框架. 已经被广泛用于集群内服务间调用. 为了大规模流量和避免单点故障, 所以服务往往是部署多实例的, 于是负载均衡就是硬需求了.

https://blog.cong.moe/post/2021-03-06-grpc-go-discovery-lb/

**Comment by SocketNet** [comment]
看起来挺难用的

**Reply by zcong1993** [reply]
使用起来还行, 扩展和学习原理比较难. 文章刚更新了一版.

**Comment by virus20200317** [comment]
写的挺好的 k8s 希望resolver里加个 k8s， 毕竟用的人还挺多的

**Comment by wuranxu** [comment]
大佬，我现在也是用的etcd+robin的方式，代码大体上差不多，也是watch一个服务的key，因为我是go网关+Python服务端，我最近压测发现: 发现经过lb以后会经常报dial context deadline的超时问题。但是当我放弃lb，指定对应的host+port的时候就没再出现连接问题了。
一直也找不到解决方案，不知道是不是lb写的有问题，所以想问下大佬有没有对类似的lb的接口压测过呢，会否也遇到类似的问题。

**Reply by wuranxu** [reply]
grpc的连接我是用的dial+close，每次都会开辟一个连接，然后我发现resolver的build方法每次都会执行，会出现很多goroutine在watch，这里我稍微改动了一下，但是还是没有效果。。枯了

**Reply by zcong1993** [reply]
生产直接用 etcd 自带的吧, https://github.com/etcd-io/etcd/tree/main/client/v3/naming

**Reply by wuranxu** [reply]
> 生产直接用 etcd 自带的吧, https://github.com/etcd-io/etcd/tree/main/client/v3/naming

我看了doc.go里面的demo，试用了一下，确实没再报连接超时的问题了~多谢大佬~回头再研究下写的有啥问题 👍
